File: problems/0016-3sum-closest/solution.py
"""
Problem Number: 16
Problem Name: 3Sum Closest
Difficulty: Medium
Topics: Array, Two Pointers, Sorting
Companies: Facebook, Amazon, Google
"""

from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        """
        Approach: Sorting + Two Pointers
        Time Complexity: O(n²)
        Space Complexity: O(1)
        """
        nums.sort()
        closest_sum = float('inf')
        
        # Initialize with first three elements as a baseline (handling infinite properly requires valid init)
        if len(nums) >= 3:
            closest_sum = nums[0] + nums[1] + nums[2]
            
        for i in range(len(nums) - 2):
            l, r = i + 1, len(nums) - 1
            while l < r:
                current_sum = nums[i] + nums[l] + nums[r]
                
                if abs(target - current_sum) < abs(target - closest_sum):
                    closest_sum = current_sum
                    
                if current_sum == target:
                    return current_sum
                elif current_sum < target:
                    l += 1
                else:
                    r -= 1
        
        return closest_sum

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.threeSumClosest([-1,2,1,-4], 1)) # 2
    print(solution.threeSumClosest([0,0,0], 1))     # 0

File: problems/0016-3sum-closest/solution.js
/**
 * @problemNumber: 16
 * @problemName: 3Sum Closest
 * @difficulty: Medium
 * @topics: Array, Two Pointers, Sorting
 * @companies: Facebook, Amazon, Google
 */

/**
 * Approach: Sorting + Two Pointers
 * Time Complexity: O(n²)
 * Space Complexity: O(log n) to O(n)
 *
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
function threeSumClosest(nums, target) {
    nums.sort((a, b) => a - b);
    let closestSum = nums[0] + nums[1] + nums[2];

    for (let i = 0; i < nums.length - 2; i++) {
        let left = i + 1;
        let right = nums.length - 1;

        while (left < right) {
            const currentSum = nums[i] + nums[left] + nums[right];

            if (Math.abs(target - currentSum) < Math.abs(target - closestSum)) {
                closestSum = currentSum;
            }

            if (currentSum === target) {
                return currentSum;
            } else if (currentSum < target) {
                left++;
            } else {
                right--;
            }
        }
    }

    return closestSum;
}

// Test cases
console.log(threeSumClosest([-1, 2, 1, -4], 1)); // 2
console.log(threeSumClosest([0, 0, 0], 1));     // 0

module.exports = { threeSumClosest };
